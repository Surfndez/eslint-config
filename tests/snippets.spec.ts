import { readdirSync, Dirent } from "fs"
import { join, relative } from "path"
import { CLIEngine } from "eslint"
import myConfig from "../src/index"

/**
 * gets snippet files
 * @param dir the path to get all snippets in recursively.
 * @param basePath Specify a basepath if you want the returned paths to be relative to this path
 */
function getSnippetFiles(
  dir: string = __dirname,
  basePath: string = ""
): string[] {
  const entries: Dirent[] = readdirSync(dir, { withFileTypes: true })
  let snippetPaths: string[] = []
  for (const entry of entries) {
    if (entry.isFile()) {
      const fullPath = join(dir, entry.name)
      const relativePath = basePath ? relative(basePath, fullPath) : fullPath
      snippetPaths.push(relativePath)
    } else if (entry.isDirectory()) {
      const kids: string[] = getSnippetFiles(join(dir, entry.name), basePath)
      snippetPaths = snippetPaths.concat(kids)
    } else {
      //console.warn("unexpected file type in snapshots:", entry)
    }
  }
  return snippetPaths
}

const snippetBase = join(__dirname, "..", "test-data", "snippets")

test.each(getSnippetFiles(join(snippetBase, "should-error"), snippetBase))(
  "should error %s",
  snippetFile => {
    const result = lint(snippetFile)
    expect(result.errorCount).toBeGreaterThan(0)
    expect(result.warningCount).toEqual(0)
    expect(result).toMatchSnapshot(snippetFile)
  }
)

test.each(getSnippetFiles(join(snippetBase, "should-warn"), snippetBase))(
  "should error %s",
  snippetFile => {
    const result = lint(snippetFile)
    expect(result.errorCount).toEqual(0)
    expect(result.warningCount).toBeGreaterThan(0)
    expect(result).toMatchSnapshot(snippetFile)
  }
)

test.each(getSnippetFiles(join(snippetBase, "should-pass"), snippetBase))(
  "should error %s",
  snippetFile => {
    const result = lint(snippetFile)
    // result.results.forEach(r => console.log("result:", r))
    expect(result.errorCount).toEqual(0)
    expect(result.warningCount).toEqual(0)
    expect(result).toMatchSnapshot(snippetFile)
  }
)

function lint(snippetFile): CLIEngine.LintReport {
  const cli = new CLIEngine({
    useEslintrc: false,
    ignore: false,
    baseConfig: myConfig
  })
  const fullPath = join(snippetBase, snippetFile)
  const report = cli.executeOnFiles([fullPath])
  return removeFilePaths(report)
}

function removeFilePaths(report: CLIEngine.LintReport): CLIEngine.LintReport {
  report.results.forEach(p => (p.filePath = relative(snippetBase, p.filePath)))
  return report
}
