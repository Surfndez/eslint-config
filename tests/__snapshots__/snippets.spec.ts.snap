// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should error should-error/ctor-accessibility-modifer.ts: should-error/ctor-accessibility-modifer.ts 1`] = `
Object {
  "errorCount": 2,
  "fixableErrorCount": 0,
  "fixableWarningCount": 0,
  "results": Array [
    Object {
      "errorCount": 2,
      "filePath": "/Users/scott/src/activescott-eslint-config/test-data/snippets/should-error/ctor-accessibility-modifer.ts",
      "fixableErrorCount": 0,
      "fixableWarningCount": 0,
      "messages": Array [
        Object {
          "column": 3,
          "endColumn": 19,
          "endLine": 2,
          "line": 2,
          "message": "Missing accessibility modifier on method definition constructor.",
          "messageId": "missingAccessibility",
          "nodeType": "MethodDefinition",
          "ruleId": "@typescript-eslint/explicit-member-accessibility",
          "severity": 2,
        },
        Object {
          "column": 17,
          "line": 2,
          "message": "Unexpected empty constructor.",
          "messageId": "unexpected",
          "nodeType": "FunctionExpression",
          "ruleId": "@typescript-eslint/no-empty-function",
          "severity": 2,
        },
      ],
      "source": "export class TestClass {
  constructor() {}
}
",
      "warningCount": 0,
    },
  ],
  "usedDeprecatedRules": Array [],
  "warningCount": 0,
}
`;

exports[`should error should-error/ctor-fields-accessibility-modifer.ts: should-error/ctor-fields-accessibility-modifer.ts 1`] = `
Object {
  "errorCount": 1,
  "fixableErrorCount": 0,
  "fixableWarningCount": 0,
  "results": Array [
    Object {
      "errorCount": 1,
      "filePath": "/Users/scott/src/activescott-eslint-config/test-data/snippets/should-error/ctor-fields-accessibility-modifer.ts",
      "fixableErrorCount": 0,
      "fixableWarningCount": 0,
      "messages": Array [
        Object {
          "column": 22,
          "endColumn": 48,
          "endLine": 2,
          "line": 2,
          "message": "Missing accessibility modifier on parameter property someField.",
          "messageId": "missingAccessibility",
          "nodeType": "TSParameterProperty",
          "ruleId": "@typescript-eslint/explicit-member-accessibility",
          "severity": 2,
        },
      ],
      "source": "export class TestClass {
  public constructor(readonly someField: string) {}
}
",
      "warningCount": 0,
    },
  ],
  "usedDeprecatedRules": Array [],
  "warningCount": 0,
}
`;

exports[`should error should-error/semi.ts: should-error/semi.ts 1`] = `
Object {
  "errorCount": 1,
  "fixableErrorCount": 1,
  "fixableWarningCount": 0,
  "results": Array [
    Object {
      "errorCount": 1,
      "filePath": "/Users/scott/src/activescott-eslint-config/test-data/snippets/should-error/semi.ts",
      "fixableErrorCount": 1,
      "fixableWarningCount": 0,
      "messages": Array [
        Object {
          "column": 27,
          "fix": Object {
            "range": Array [
              56,
              58,
            ],
            "text": ")",
          },
          "line": 3,
          "message": "Extra semicolon.",
          "nodeType": "ExpressionStatement",
          "ruleId": "semi",
          "severity": 2,
        },
      ],
      "source": "import process from \\"process\\"

process.stdout.write(\\"hi\\");
",
      "warningCount": 0,
    },
  ],
  "usedDeprecatedRules": Array [],
  "warningCount": 0,
}
`;

exports[`should error should-pass/ctor-accessibility-modifer.ts: should-pass/ctor-accessibility-modifer.ts 1`] = `
Object {
  "errorCount": 0,
  "fixableErrorCount": 0,
  "fixableWarningCount": 0,
  "results": Array [
    Object {
      "errorCount": 0,
      "filePath": "/Users/scott/src/activescott-eslint-config/test-data/snippets/should-pass/ctor-accessibility-modifer.ts",
      "fixableErrorCount": 0,
      "fixableWarningCount": 0,
      "messages": Array [],
      "warningCount": 0,
    },
  ],
  "usedDeprecatedRules": Array [],
  "warningCount": 0,
}
`;

exports[`should error should-pass/ctor-fields-accessibility-modifer.ts: should-pass/ctor-fields-accessibility-modifer.ts 1`] = `
Object {
  "errorCount": 0,
  "fixableErrorCount": 0,
  "fixableWarningCount": 0,
  "results": Array [
    Object {
      "errorCount": 0,
      "filePath": "/Users/scott/src/activescott-eslint-config/test-data/snippets/should-pass/ctor-fields-accessibility-modifer.ts",
      "fixableErrorCount": 0,
      "fixableWarningCount": 0,
      "messages": Array [],
      "warningCount": 0,
    },
  ],
  "usedDeprecatedRules": Array [],
  "warningCount": 0,
}
`;

exports[`should error should-warn/no-console.ts: should-warn/no-console.ts 1`] = `
Object {
  "errorCount": 0,
  "fixableErrorCount": 0,
  "fixableWarningCount": 0,
  "results": Array [
    Object {
      "errorCount": 0,
      "filePath": "/Users/scott/src/activescott-eslint-config/test-data/snippets/should-warn/no-console.ts",
      "fixableErrorCount": 0,
      "fixableWarningCount": 0,
      "messages": Array [
        Object {
          "column": 1,
          "endColumn": 12,
          "endLine": 1,
          "line": 1,
          "message": "Unexpected console statement.",
          "messageId": "unexpected",
          "nodeType": "MemberExpression",
          "ruleId": "no-console",
          "severity": 1,
        },
      ],
      "source": "console.log(\\"hi\\")
",
      "warningCount": 1,
    },
  ],
  "usedDeprecatedRules": Array [],
  "warningCount": 1,
}
`;
